### PR TITLE
[CI] Quantization workflow

### PR DESCRIPTION
**Pull request by SunMarc**
# What does this PR do ? 
This PR adds a workflow for quantization tests + related dockerfile. Since we merged the [HfQuantizer PR](https://github.com/huggingface/transformers/pull/26610), the community started integrating their own quantizers into transformers (e.g. [AQML](https://github.com/huggingface/transformers/pull/28928) and much more in the future). This will lead to many third party libraries in the Dockerfile `huggingface/transformers-all-latest-gpu`. To limit the impact of these libraries on transformers tests, I propose to create a seperate dockerfile + workflow.

### CODE REVIEW

**Review by younesbelkada — COMMENTED** [review]
Thank you very much ! Can you try to trigger a run of the quantization slow tests and make sure we get the slack notifications (after building the quantization docker image) ? 🙏 
For that you can comment out all jobs except the quantization tests job in `self-scheduled.yml` & `build-docker.yml` and modify the run condition on on push to the name of the branch you are working on right now (there might be a better way to trigger only the quantization tests but all tests are on the same workflow file)

File: .github/workflows/self-scheduled.yml
Comment (younesbelkada) [inline .github/workflows/self-scheduled.yml:318]:
```suggestion
      - uses: actions/checkout@v3
```
I think this can be completely removed in favor of uses: actions/checkout@v3: https://github.com/actions/checkout

See e.g.: https://github.com/huggingface/trl/blob/main/.github/workflows/slow-tests.yml#L33C15-L33C31

**Review by younesbelkada — COMMENTED** [review]
I left a comment ! (this can be applied to all our workflow file that used that logic for checking out to the current main branch)

**Comment by SunMarc** [comment]
> For that you can comment out all jobs except the quantization tests job in self-scheduled.yml & build-docker.yml and modify the run condition on on push to the name of the branch you are working on right now (there might be a better way to trigger only the quantization tests but all tests are on the same workflow file)

I tried to build the docker image but it gives the following error: 

`ERROR: Could not install packages due to an OSError: [Errno 28] No space left on device`.
Do you know how to fix that ? 

Also, I see that  `transformers-all-latest-gpu` docker image is not being updated for the last two days since the [installation](https://github.com/huggingface/transformers/actions/runs/7924158967/job/21635225922) fails because of aqml library that requires python 3.10 at least and we uses 3.8 for now. We will have to change that in the quantization dockerfile.

File: .github/workflows/self-scheduled.yml
Comment (SunMarc) [reply .github/workflows/self-scheduled.yml:318]: Thank you, I will first test it without and try to add that when I'm able to run the workflow successfully !

**Comment by BlackSamorez** [comment]
The only reason `aqlm` requires `python>=3.10` is a single `match-case` statement in a non-critical place.

I was able to run `aqlm` on `python 3.8` no problem. I replace the statement with an `if-else` statement and lower the requirement if necessary.

**Comment by younesbelkada** [comment]
@SunMarc thanks! 
in trl i build a docker image with python 3.10: https://github.com/huggingface/trl/blob/main/docker/trl-source-gpu/Dockerfile maybe you can take some inspiration from that dockerfile? 🙏 I am not sure why you are getting that error currently as the commands looks correct. 
@BlackSamorez yes that would be great if you can also support python 3.8 for AQLM 🙏 Thanks !

**Comment by SunMarc** [comment]
> I was able to run aqlm on python 3.8 no problem otherwise. I can replace the statement with an if-else statement and lower the requirement if necessary.

Yes that would be for the best ! I prefer to keep running the quantization tests with python 3.8 since this is what we are actually doing for all transformers tests ! Moreover, it will be better for the users since we keep the requirement low. LMK when it is done ! @BlackSamorez 
Otherwise, I will modify the dockerfile and use conda to install python 3.10 as suggested by @younesbelkada.

**Comment by BlackSamorez** [comment]
@SunMarc `aqlm` will support python `>=3.8` starting version `1.0.2`. I'm [1 PR](https://github.com/Vahe1994/AQLM/pull/26) away from releasing it.

**Comment by SunMarc** [comment]
Perfect ! I will wait for your PR to be merged then + release then if it doesn't take too much time. Please keep me updated ! Otherwise, I can first merge this PR without aqlm and add it afterwards.

**Comment by BlackSamorez** [comment]
@SunMarc `aqlm==1.0.2` is out. May I ask you to please update the docker images?

**Comment by SunMarc** [comment]
I was able to build the image but I don't have the permission to push the image cc @ydshieh 

`#22 ERROR: failed to push huggingface/transformers-quantization-latest-gpu: push access denied, repository does not exist or may require authorization: server message: insufficient_scope: authorization failed`

**Comment by ydshieh** [comment]
@SunMarc  Do you builded/pushed via transformers' github actions? If so, do you have a job run link?

**Comment by SunMarc** [comment]
Yes ! Here's the link to the [job] (https://github.com/huggingface/transformers/actions/runs/7979164961)

**Comment by younesbelkada** [comment]
it's quite strange because the workflow was indeed able to login: https://github.com/huggingface/transformers/actions/runs/7979164961/job/21802381289#step:5:1 but fails to push ...

**Comment by younesbelkada** [comment]
In TRL and PEFT I can confirm the build & push works fine :  https://github.com/huggingface/peft/actions/workflows/build_docker_images.yml / https://github.com/huggingface/trl/actions/workflows/docker-build.yml so it's not a token issue as we use the same except if the token has expired for transformers cc @glegendre01

**Comment by ydshieh** [comment]
Hi @SunMarc it's because there is some change in the infra team on docker hub stuff.

The repository `huggingface/transformers-quantization-latest-gpu` has to be created on docker Hub first, then only after this, you can push to it.

I will ask for it.

**Comment by ydshieh** [comment]
BTW, I will review this PR tomorrow or Friday 🙏

File: .github/workflows/self-scheduled.yml
Comment (younesbelkada) [inline .github/workflows/self-scheduled.yml:336]:
```suggestion
```

File: .github/workflows/self-scheduled.yml
Comment (younesbelkada) [reply .github/workflows/self-scheduled.yml:318]: You also need to call `pip install -e .` at the begining of the job after the actions/checkout

Comment (SunMarc) [reply .github/workflows/self-scheduled.yml:318]: I will leave it as it is since I think that @ydshieh prefer not to use `actions/checkout@v3`. We can still make the change in this [PR](https://github.com/huggingface/transformers/pull/29197) if we see that it works !

File: docker/transformers-all-latest-gpu/Dockerfile
Comment (ydshieh) [inline docker/transformers-all-latest-gpu/Dockerfile:53]: Do we count this as `non quantization` tests?

File: docker/transformers-quantization-latest-gpu/Dockerfile
Comment (ydshieh) [inline docker/transformers-quantization-latest-gpu/Dockerfile:12]: Let's aim to use `torch 2.2`. See #29208

File: docker/transformers-quantization-latest-gpu/Dockerfile
Comment (ydshieh) [inline docker/transformers-quantization-latest-gpu/Dockerfile:27]: make sure at the end, we do get `torch 2.2`. See my comment in the change in #29208. Otherwise, move this after `./transformers[dev]` step.

File: docker/transformers-quantization-latest-gpu/Dockerfile
Comment (ydshieh) [inline docker/transformers-quantization-latest-gpu/Dockerfile:31]: we can also uninstall tensorflow. Or just use `./transformers[torch]` above (but we might need to install other stuffs manually, not sure)

(not a big deal, we can keep it if you are busy)

File: .github/workflows/build-docker-images.yml
Comment (ydshieh) [inline .github/workflows/build-docker-images.yml:315]: let's from now on move this to `[intel-cpu, 8-cpu, ci]` (our own VM, bigger RAM and disk)

File: .github/workflows/build-docker-images.yml
Comment (ydshieh) [inline .github/workflows/build-docker-images.yml:326]: we can remove this now I guess (once we use `[intel-cpu, 8-cpu, ci]`)

File: .github/workflows/build-docker-images.yml
Comment (ydshieh) [inline .github/workflows/build-docker-images.yml:314]: Just want to know if we intent to run `quantization` tests in a daily basis, or if we prefer to run them on each commit merged into `main`?

File: .github/workflows/self-scheduled.yml
Comment (ydshieh) [inline .github/workflows/self-scheduled.yml:301]: I am not sure if we intend to keep this in this workflow file. Isn't the goal to move this outside to a separate workflow file ?

cc @younesbelkada

File: .github/workflows/self-scheduled.yml
Comment (ydshieh) [reply .github/workflows/self-scheduled.yml:318]: yes, I talked to @younesbelkada . Il fait n'importe de quoi 😆  (joking)

File: utils/notification_service.py
Comment (ydshieh) [inline utils/notification_service.py:1046]: should be revised if we decide to move quantization tests outside this workflow file.

**Review by ydshieh — COMMENTED** [review]
Thank  you @SunMarc (and @younesbelkada ) for this initiative.

I love this idea - quantization tests become more and more, and failing on them becomes hard to track in the whole daily CI.

I have some suggestions and questions for the current version though.

File: .github/workflows/self-scheduled.yml
Comment (younesbelkada) [reply .github/workflows/self-scheduled.yml:318]: hahaha

File: .github/workflows/self-scheduled.yml
Comment (younesbelkada) [reply .github/workflows/self-scheduled.yml:301]: I think we can keep it there for now and I'll move it outside in my PR

File: docker/transformers-all-latest-gpu/Dockerfile
Comment (younesbelkada) [reply docker/transformers-all-latest-gpu/Dockerfile:53]: I think we need to keep that as optimum is a hard requirement to use GPTQ

**Comment by ydshieh** [comment]
@SunMarc I am going to merge #29208. Once done, there might be a conflict with this PR, but it should be easy to resolve. Ping me otherwise.

File: .github/workflows/build-docker-images.yml
Comment (SunMarc) [reply .github/workflows/build-docker-images.yml:314]: I think that it is better to run them on a daily basis. There are many slows tests too + I think that the quantization tests are not easily broken by changes in transformers. The bigger issue is the third party libraries that might introduce breaking changes. cc @younesbelkada

File: docker/transformers-quantization-latest-gpu/Dockerfile
Comment (SunMarc) [reply docker/transformers-quantization-latest-gpu/Dockerfile:31]: I can switch to transformers[dev-torch] if this makes more sense.

File: .github/workflows/build-docker-images.yml
Comment (SunMarc) [reply .github/workflows/build-docker-images.yml:305]: done !

File: .github/workflows/build-docker-images.yml
Comment (SunMarc) [reply .github/workflows/build-docker-images.yml:316]: removed !

**Comment by SunMarc** [comment]
I've addressed all comments. I think that we are good to merge after I checked that the docker image is built correctly with torch 2.2.0 (currently facing some issues with login into dockerhub) cc @ydshieh

File: docker/transformers-quantization-latest-gpu/Dockerfile
Comment (SunMarc) [reply docker/transformers-quantization-latest-gpu/Dockerfile:27]: Done ! I didn't have to move it after  `./transformers[dev]` step.

File: .github/workflows/build-docker-images.yml
Comment (younesbelkada) [reply .github/workflows/build-docker-images.yml:304]: Yes daily basis sounds great!

File: .github/workflows/self-scheduled.yml
Comment (ArthurZucker) [inline .github/workflows/self-scheduled.yml:318]: @ydshieh I do not understand why we don't:
1. Install from main on the docker build with `pip install git+https://github.com/hugginface/transformers@main` (and get all the necessary dependencies)
2. `pip uninstall -y transformers` in the docker 
3. `actions/checkout@v3`
4. `pip install -e  .`
(As discussed offline with @younesbelkada this should come in a separate PR)

**Review by ArthurZucker — APPROVED** [review]
Preemptively approving, LGTM! Thanks @SunMarc for splitting the workflows!

File: .github/workflows/self-scheduled.yml
Comment (ydshieh) [reply .github/workflows/self-scheduled.yml:318]: There is no deep reason for what we are doing currently: we just install `transformers` during docker image build and reuse it.

What you suggest (by installing `main` during docker image build) has a consequence that `we can't do experiment easily in a branch when we change `setup.py`).

Also by using `actions/checkout@v3`, the path to `transformers` will be different from the current approach, and therefore we have to change all the working directory path and artifact path, otherwise the report will be empty.

File: .github/workflows/build-docker-images.yml
Comment (ydshieh) [inline .github/workflows/build-docker-images.yml:299]: This is irrelevant to this PR. We have (or at least had) issue of building AMD CI images and we haven't really make progress on it. 

Better to keep this as it is in this PR.

File: .github/workflows/build-docker-images.yml
Comment (ydshieh) [reply .github/workflows/build-docker-images.yml:304]: OK

**Review by ydshieh — APPROVED** [review]
Thank you for this work! I could RIP with the daily CI now.

**Comment by ydshieh** [comment]
I would love to see the newly added `job` could be run successfully (having failing tests are fine), but I don't enforce it here.

**Pull request merged**
